### PR TITLE
Fix bug with rco camera loading in UDA camera data loader

### PR DESCRIPTION
**Pull request by samueljackson92**
This PR adds better edge case handling for UDA camera data and fixes the bug with loading RCO data.

To Test:
 - Follow the instructions in the #320 issue.

### CODE REVIEW

File: toktagger/api/core/data_loaders.py [hunk 1]
```diff
@@ -414,6 +414,9 @@ def get_sample(
             if params.frame is None:
                 params.frame = 0  # Default to first frame if not specified
 
+            print(
```
Comment (abdullah-ukaea) [inline toktagger/api/core/data_loaders.py]: can we remove this print statement now? I assume it was here for debugging purposes?

File: toktagger/api/core/data_loaders.py [hunk 2]
```diff
@@ -423,6 +426,26 @@ def get_sample(
...
+            # Convert uint16 to uint8 for Pillow compatibility (Pillow doesn't support u2)
+            if image_array.dtype == np.uint16:
+                if np.any(image_array > 255):
+                    val_range = image_array.max() - image_array.min()
+                    image_array = image_array - image_array.min()
+                    if val_range:
+                        image_array = image_array / val_range
+                    image_array = (image_array * 255).astype(np.uint8)
+                else:
+                    image_array = image_array.astype(np.uint8)
```
Comment (abdullah-ukaea) [inline toktagger/api/core/data_loaders.py]: This rescales every frame using that frame’s darkest and brightest pixels, so brightness may jump between frames. A frame containing one constant value above 255 would also become completely black. Could we instead use the camera’s declared bit depth for a consistent conversion for RCO, simply clip to 0–255 and convert to uint8?

I think the uda-xarray includes the camera’s bit depth in the data attributes
bit_depth = signal["data"].attrs["depth"]

For shot 54339, RCO reported depth = 8, even though NumPy stores the returned array as uint16. So the code can use that metadata to perform a consistent conversion to uint8.

**Review by abdullah-ukaea — COMMENTED** [review]
Looks good, please address the 2 comments in the PR.

I also looked at  tests/api/core/test_data_loaders.py and would like to request one more change.

Could we add or extend a data-loader test to cover this regression? The existing UDA camera test only covers RBA’s 2-D uint8 output, whereas issue https://github.com/ukaea/toktagger/issues/320 was caused by RCO returning (1, height, width, 3) uint16 data.

**Comment by abdullah-ukaea** [comment]
Also worth documenting to anyone testing this PR that there is a known bug if you click next very quickly it will break the video view (I will document this in a formal issue when this is merged into dev) that will be fixed in a follow up PR that is caused by the latency of this specific UDA camera source RCO that is not found on rba camera.

We have ideas on how to fix this.

File: toktagger/api/core/data_loaders.py (hunk 2 above)
Comment (samueljackson92) [reply toktagger/api/core/data_loaders.py]: Yes good point.

File: tests/api/core/test_data_loaders.py [hunk 3]
```diff
@@ -165,6 +166,106 @@ def test_uda_camera_loader(uda_env_vars):
...
+        shot_id=54339,
+        data=uda_shot,
+        _id="test",
+        project_id="test",
+        validated_annotations=False,
+    )
+
+    raw_values = numpy.array([[0, 128], [200, 255]], dtype=numpy.uint16)
+    fake_dataset = xarray.Dataset(
+        {"data": xarray.DataArray(raw_values, dims=["x", "y"], attrs={"depth": 8})}
```
Comment (abdullah-ukaea) [inline tests/api/core/test_data_loaders.py:186]: These tests only use 2-D images, but the original RCO error (documented in https://github.com/ukaea/toktagger/issues/320) happened with a three-channel uint16 image I think. That means the tests check the scaling change, but not the exact image format that previously failed.

Could we add one test using an RCO-like (1, height, width, 3) uint16 image with depth=8,and check that the returned PNG is a (height, width, 3) uint8 image?

**Review by abdullah-ukaea — COMMENTED** [review]
The data-loader changes look good to me. I just have one final comment on the new tests:

Could we add a test for the three-channel uint16 image shape that caused the original RCO loading problem? 
@samueljackson92

File: tests/api/core/test_data_loaders.py (hunk 3 above)
Comment (abdullah-ukaea) [reply tests/api/core/test_data_loaders.py:186]: Here is an example of what I mean.

It mirrors the original RCO format: a leading time dimension plus RGB `uint16` pixels with `depth=8`.

```python
def test_uda_camera_loader_rco_rgb_uint16(monkeypatch):
    # RCO returns one time frame of three-channel uint16 image data.
    raw_values = numpy.array(
        [
            [
                [[0, 64, 128], [255, 200, 100]],
                [[10, 20, 30], [40, 50, 60]],
            ]
        ],
        dtype=numpy.uint16,
    )
    fake_dataset = xarray.Dataset(
        {
            "data": xarray.DataArray(
                raw_values,
                dims=["time", "height", "width", "channel"],
                attrs={"depth": 8},
            )
        }
    )
    monkeypatch.setattr(
        data_loaders.xr, "open_dataset", lambda *args, **kwargs: fake_dataset
    )

    sample = Sample(
        shot_id=54339,
        data=ShotData(protocol="uda", signal_names=["rco"]),
        _id="test",
        project_id="test",
        validated_annotations=False,
    )

    image_data = data_loaders.UDACameraDataLoader().get_sample(
        sample, params=ImageParams(name="image", frame=0)
    )
    result = numpy.array(Image.open(io.BytesIO(base64.b64decode(image_data.values))))

    assert result.shape == (2, 2, 3)
    assert result.dtype == numpy.uint8
    assert numpy.array_equal(result, raw_values.squeeze(axis=0).astype(numpy.uint8))
```

**Comment by samueljackson92** [comment]
Updated.

**Comment by abdullah-ukaea** [comment]
LGTM, this is merge ready now

**Review by abdullah-ukaea — APPROVED** [review]
